sdk/python/kfp/dsl/metrics_utils.py

```diff
@@ -0,0 +1,129 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from kfp.dsl import artifact_utils
+from typing import Any, List
+
+class ComplexMetricsBase(object):
+
+  def get_schema(self):
+    return self._schema
+
+  def get_metrics(self):
+    artifact_utils.verify_schema_instance(self._schema, self._values)
+    return self._values
+
+  def __init__(self, schema_file: str):
+    self._schema = artifact_utils.read_schema_file(schema_file)
+
+    self._type_name, self._metric_fields = artifact_utils.parse_schema(
+      self._schema)
+
+    self._values = {}
+
+
+class ConfidenceMetrics(ComplexMetricsBase):
+  """Metrics class representing a Confidence Metrics."""
+
+    # Initialization flag to support setattr / getattr behavior.
+  _initialized = False
+
+  def __getattr__(self, name: str) -> Any:
+    """Custom __getattr__ to allow access to metrics schema fields."""
+
+    if name not in self._metric_fields:
+      raise AttributeError('No field: {} in metrics.'.format(name))
+
+    return self._values[name]
+
+  def __setattr__(self, name: str, value: Any):
+    """Custom __setattr__ to allow access to metrics schema fields."""
+
+    if not self._initialized:
```

Comment:
This is required because in the init various other instance members are set which also internally calls the __setattr__ and __getattr__. Once the instance is initialized the objects only support setting specific attributes.

Reply:
Thanks for the explanation.